File: src/p1019/python/solution.py
class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        result, stk = [], []
        p = head
        while p:
            result.append(p.val)
            p = p.next
        for i in range(len(result)):
            while len(stk) and result[stk[-1]] < result[i]:
                result[stk.pop()] = result[i]
            stk.append(i)
        while len(stk):
            result[stk.pop()] = 0
        return result

File: src/p1019/cpp/solution.cpp
class Solution {
public:
    vector<int> nextLargerNodes(ListNode *head) {
        vector<int> result;
        for (auto p = head; p; p = p->next) {
            result.push_back(p->val);
        }
        stack<size_t> stk;
        for (size_t i = 0; i < result.size(); i++) {
            while (!stk.empty() && result[stk.top()] < result[i]) {
                result[stk.top()] = result[i]; stk.pop();
            }
            stk.push(i);
        }
        while (!stk.empty()) { result[stk.top()] = 0; stk.pop(); }
        return result;
    }
};
